### PR TITLE
Deliverable D: Review for sub-module creation and code for output file format. GROUP D

### PR DESCRIPTION
**Pull request by revanth-sridhar**
Hi team,

Could you please review this pull request for, 
- adding cadmium and DESTimes as sub modules, 
- handling the duplicate NDTime.h file in the project folder.
- creating a new .csv file by formatting the output file as per the requirement. 

Many thanks in advance.

### CODE REVIEW

File: src/transform.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+/**
+* Organization    : ARSLab - Carleton University
+* Original author : Cristina Ruiz Martin
+* Modified by     : Manoj Goli and Revanth Sridhar
+*/
+
+#include<string.h>
+#include<stdio.h>
+
```
Comment (shubhamthakur463) [inline src/transform.cpp]: <string.h> and <stdio.h> are not needed here, as they are already included inside transform.hpp
Kindly check the repetition of header files in transform.hpp and transform.cpp

**Comment by gagandeep8030** [comment]
Please update README.md file to include transform.hpp .

File: README.md [hunk 2]
```diff
@@ -42,9 +42,7 @@ This ABP simulator is implemented in Cadmium. The behaviour and specification of
 ### lib [This folder contains the third party library files]
 1. cadmium [This folder contains the cadmium library files]
 2. DESTimes [This folder contains the DESTimes library files]
-3. vendor [This folder contains vendor header files]
-	- iestream.hpp
-	- NDTime.hpp
+3. iestream.hpp
 	
 ### src [This folder contains the source files used for the simualator]
```
Comment (shubhamthakur463) [inline README.md:49]: Kindly update the folder structure in README.md to add transform.cpp inside the src folder.

**Comment by gagandeep8030** [comment]
Inside README.md , please update the folder structure to include abp_processed_output.csv inside the data/output folder.

Also, include new output files for unit tests i.e. receiver, sender and subnet in README.md

File: src/top_model/main.cpp [hunk 3]
```diff
@@ -211,5 +213,12 @@ std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> TOP = std::make_share
     r.run_until(NDTime("04:00:00:000"));
     auto elapsed = std::chrono::duration_cast<std::chrono::duration<double, std::ratio<1>>>(hclock::now() - start).count();
     cout << "Simulation took:" << elapsed << "sec" << endl;
+
+    FILE *input;
```
Comment (gagandeep8030) [inline src/top_model/main.cpp:223]: It is a suggestion for better understanding, you can add comments for file pointers and function called as this part is newly added. Same can be implemented in all main.cpp files.

**Comment by shubhamthakur463** [comment]
In README.md, after running the simulator kindly include instructions to open and view the new modified output file for ABP simulator and unit tests.

**Review by shubhamthakur463 — APPROVED** [review]
Part-D approved.

File: src/top_model/main.cpp (hunk 3 above)
Comment (manoj-goli) [reply src/top_model/main.cpp:223]: Thanks,
It has been resolved.

**Comment by manoj-goli** [comment]
> Please update README.md file to include transform.hpp .

Thanks,
It has been resolved.

**Comment by manoj-goli** [comment]
> Inside README.md , please update the folder structure to include abp_processed_output.csv inside the data/output folder.
> 
> Also, include new output files for unit tests i.e. receiver, sender and subnet in README.md

Thanks,
It has been resolved.

**Comment by manoj-goli** [comment]
> In README.md, after running the simulator kindly include instructions to open and view the new modified output file for ABP simulator and unit tests.

Thanks,
It has been resolved.

**Review by manoj-goli — APPROVED** [review]

**Review by gagandeep8030 — APPROVED** [review]
Part D approved.

File: src/transform.cpp (hunk 1 above)
Comment (revanth-sridhar) [reply src/transform.cpp]: thanks a lot for letting us know. We have fixed it.

File: include/transform.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/** \brief filter function helps in formatting the output.
+* This function takes the unformatted output files from the simulator as 
+* the input and generates formatted output for the ABP simulator and 
+* all the unit tests (sender, subnet, receiver).
+*
+* for example (formatted output), 
+* time			value	port	        component
+* 00:00:10:000	5	    out				generator_con
+* 00:00:20:000	1	    packet_sent_out	sender1
```
Comment (cruizm) [inline include/transform.hpp:9]: This is not properly aligned

Comment (revanth-sridhar) [reply include/transform.hpp:9]: Many thanks for this review. We would like to update that this was addressed. 
Same comments was identified in deliverable E and has been addressed in Deliverable E. Master has it reflected.

**Pull request merged**